src/actions/receive-profile.js

```diff
@@ -38,13 +44,116 @@ export function waitingForProfileFromAddon(): Action {
   };
 }
 
-export function viewProfile(profile: Profile): Action {
-  return {
-    type: 'VIEW_PROFILE',
-    profile,
+export function viewProfile(
+  profile: Profile,
+  pathInZipFile: ?string
+): ThunkAction<void> {
+  return (dispatch, getState) => {
+    const threadIndexes = profile.threads.map((_, threadIndex) => threadIndex);
+    let hiddenThreadIndexes;
+    let selectedThreadIndex = getSelectedThreadIndexOrNull(getState());
+
+    if (selectedThreadIndex === null) {
+      // This is a new profile, it has no selected thread index.
+      hiddenThreadIndexes = _hideIdleThreads(profile);
+    } else {
+      // Pull the existing data that was URL encoded.
+      hiddenThreadIndexes = getHiddenThreads(getState())
+        // Try to use the hidden threads specified in the URL, but ensure that the indexes
+        // are correct by filtering out invalid ones.
+        .filter(index => threadIndexes.includes(index));
+    }
+
+    if (
```

Comment:
please add an explicit condition `selectedThreadIndex === null`; this is implicitely contained in the other conditions but I find it difficult to understand it at the first read